src/alignment/pairwise.rs

```diff
@@ -96,37 +95,52 @@ macro_rules! align {
 
                     // score for deletion
-                    let d_score = max(
-                        $aligner.S[prev][$state.j] + $aligner.gap_open,
-                        $aligner.D[prev][$state.j] + $aligner.gap_extend
-                    );
+                    let d_open = $aligner.S[prev][$state.j] + $aligner.gap_open;
+                    let d_extend = $aligner.D[prev][$state.j] + $aligner.gap_extend;
+
+                    if d_open > d_extend {
+                        $aligner.D_traceback.subst($state.i, $state.j);
+                        $aligner.D[$state.col][$state.j] = d_open;
+                    } else {
+                        $aligner.D_traceback.del($state.i, $state.j);
+                        $aligner.D[$state.col][$state.j] = d_extend;
+                    }
+
                     // score for insertion
-                    let i_score = max(
-                        $aligner.S[$state.col][$state.j-1] + $aligner.gap_open,
-                        $aligner.I[$state.col][$state.j-1] + $aligner.gap_extend
-                    );
+                    let i_open = $aligner.S[$state.col][$state.j-1] + $aligner.gap_open;
+                    let i_extend = $aligner.I[$state.col][$state.j-1] + $aligner.gap_extend;
+
+                    if i_open > i_extend {
+                        $aligner.I_traceback.subst($state.i, $state.j);
+                        $aligner.I[$state.col][$state.j] = i_open;
+                    } else {
+                        $aligner.I_traceback.ins($state.i, $state.j);
+                        $aligner.I[$state.col][$state.j] = i_extend;
+                    };
+
+
                     // score for substitution
-                    $state.score = $aligner.S[prev][$state.j-1] + ($aligner.score)(a, b);
+                    let match_score = ($aligner.score)(a, b);
+                    $state.score = $aligner.S[prev][$state.j-1] + match_score;
+                    $aligner.S_traceback.subst($state.i, $state.j);
 
-                    if d_score > $state.score {
-                        $state.score = d_score;
-                        $aligner.traceback.del($state.i, $state.j);
-                    }
-                    else if i_score > $state.score {
-                        $state.score = i_score;
-                        $aligner.traceback.ins($state.i, $state.j);
+                    let from_d = $aligner.D[prev][$state.j-1] + match_score;
+                    let from_i = $aligner.I[prev][$state.j-1] + match_score;
+
+                    if from_d > $state.score {
+                        $state.score = from_d;
+                        $aligner.S_traceback.del($state.i, $state.j);
                     }
-                    else {
-                        $aligner.traceback.subst($state.i, $state.j);
+
+                    if from_i > $state.score {
+                        $state.score = from_i;
+                        $aligner.S_traceback.ins($state.i, $state.j);
```

Comment:
I think I am still missing something here. I always thought that the traceback matrix should be filled with the decision, based on the optimal score. However, while the decision is now made in line 130 and 135, you also fill the traceback matrix before (line 125, 117, 114, 105). Isn't this redundant?